# Import pandas

In [190]:
import pandas

In [191]:
selected_words = ['awesome', 'great', 'fantastic', 'amazing', 'love', 'horrible', 'bad', 'terrible', 'awful', 'wow', 'hate']

# Load product data

In [192]:
products = pandas.read_csv('amazon_baby.csv')

# View data

In [196]:
products.head()

name  \
0                           Planetwise Flannel Wipes   
1                              Planetwise Wipe Pouch   
2                Annas Dream Full Quilt with 2 Shams   
3  Stop Pacifier Sucking without tears with Thumb...   
4  Stop Pacifier Sucking without tears with Thumb...   

                                              review  rating  
0  These flannel wipes are OK, but in my opinion ...       3  
1  it came early and was not disappointed. i love...       5  
2  Very soft and comfortable and warmer than it l...       5  
3  This is a product well worth the purchase.  I ...       5  
4  All of my kids have cried non-stop when I trie...       5

In [197]:
len(products)

183531

In [189]:
products.describe()

rating
count  183531.000000
mean        4.120448
std         1.285017
min         1.000000
25%         4.000000
50%         5.000000
75%         5.000000
max         5.000000

In [198]:
products.tail()

name  \
183526  Baby Teething Necklace for Mom Pretty Donut Sh...   
183527  Baby Teething Necklace for Mom Pretty Donut Sh...   
183528   Abstract 2 PK Baby / Toddler Training Cup (Pink)   
183529  Baby Food Freezer Tray - Bacteria Resistant, B...   
183530  Best 2 Pack Baby Car Shade for Kids - Window S...   

                                                   review  rating  
183526  Such a great idea! very handy to have and look...       5  
183527  This product rocks!  It is a great blend of fu...       5  
183528  This item looks great and cool for my kids.......       5  
183529  I am extremely happy with this product. I have...       5  
183530  I love this product very mush . I have bought ...       5

# Split review string

In [199]:
products['review_split'] = products['review'].str.lower().str.split()

In [140]:
products['review_split']

0         [these, flannel, wipes, are, ok,, but, in, my,...
1         [it, came, early, and, was, not, disappointed....
2         [very, soft, and, comfortable, and, warmer, th...
3         [this, is, a, product, well, worth, the, purch...
4         [all, of, my, kids, have, cried, non-stop, whe...
5         [when, the, binky, fairy, came, to, our, house...
6         [lovely, book,, it's, bound, tightly, so, you,...
7         [perfect, for, new, parents., we, were, able, ...
8         [a, friend, of, mine, pinned, this, product, o...
9         [this, has, been, an, easy, way, for, my, nann...
10        [i, love, this, journal, and, our, nanny, uses...
11        [this, book, is, perfect!, i'm, a, first, time...
12        [i, originally, just, gave, the, nanny, a, pad...
13        [i, thought, keeping, a, simple, handwritten, ...
14        [space, for, monthly, photos,, info, and, a, l...
15        [i, bought, this, calender, for, myself, for, ...
16        [i, love, this, little, calend

# Add word_count column

In [210]:
def get_word_count_dict(row):
    my_dict = {}
    if type(row) is float:
        my_dict = {}
    else:
        for item in row:
            my_dict[item] = item.count(item)
    return my_dict

In [211]:
get_word_count_dict(['foo', 'bar', 'bax'])

{'bar': 1, 'bax': 1, 'foo': 1}

In [212]:
products['word_count'] = products['review_split'].apply(lambda x: get_word_count_dict(x))

# Add selected_word columns

In [216]:
def create_word_col(word_count, word):
    count = 0
    if type(word_count) is not float:
        if word in word_count:
            return word_count[word]
        else:
            return 0

In [217]:
for word in selected_words:
    products[word] = products['word_count'].apply(lambda x: create_word_col(x, word))

In [218]:
products.head()

name  \
0                           Planetwise Flannel Wipes   
1                              Planetwise Wipe Pouch   
2                Annas Dream Full Quilt with 2 Shams   
3  Stop Pacifier Sucking without tears with Thumb...   
4  Stop Pacifier Sucking without tears with Thumb...   

                                              review  rating  \
0  These flannel wipes are OK, but in my opinion ...       3   
1  it came early and was not disappointed. i love...       5   
2  Very soft and comfortable and warmer than it l...       5   
3  This is a product well worth the purchase.  I ...       5   
4  All of my kids have cried non-stop when I trie...       5   

                                        review_split  \
0  [these, flannel, wipes, are, ok,, but, in, my,...   
1  [it, came, early, and, was, not, disappointed....   
2  [very, soft, and, comfortable, and, warmer, th...   
3  [this, is, a, product, well, worth, the, purch...   
4  [all, of, my, kids, have, cried, non-stop, whe...   

                                          word_count  awesome  great  \
0  {u'and': 1, u'wipes,': 1, u'because': 1, u'ord...        0      0   
1  {u'and': 1, u'love': 1, u'it': 1, u'highly': 1...        0      0   
2  {u'and': 1, u'quilt': 1, u'it': 1, u'comfortab...        0      0   
3  {u'and': 1, u'ingenious': 1, u'love': 1, u'pos...        0      0   
4  {u'and': 1, u'parents!!': 1, u'all': 1, u'help...        0      1   

   fantastic  amazing  love  horrible  bad  terrible  awful  wow  hate  
0          0        0     0         0    0         0      0    0     0  
1          0        0     1         0    0         0      0    0     0  
2          0        0     0         0    0         0      0    0     0  
3          0        0     1         0    0         0      0    0     0  
4          0        0     0         0    0         0      0    0     0

# Add awesome_count for validation

In [219]:
def awesome_count(word_count):
    if 'awesome' in word_count:
        return word_count['awesome']
    else:
        return 0

In [220]:
products['awesome_count'] = products['word_count'].apply(lambda x: awesome_count(x))

In [221]:
products.head()

name  \
0                           Planetwise Flannel Wipes   
1                              Planetwise Wipe Pouch   
2                Annas Dream Full Quilt with 2 Shams   
3  Stop Pacifier Sucking without tears with Thumb...   
4  Stop Pacifier Sucking without tears with Thumb...   

                                              review  rating  \
0  These flannel wipes are OK, but in my opinion ...       3   
1  it came early and was not disappointed. i love...       5   
2  Very soft and comfortable and warmer than it l...       5   
3  This is a product well worth the purchase.  I ...       5   
4  All of my kids have cried non-stop when I trie...       5   

                                        review_split  \
0  [these, flannel, wipes, are, ok,, but, in, my,...   
1  [it, came, early, and, was, not, disappointed....   
2  [very, soft, and, comfortable, and, warmer, th...   
3  [this, is, a, product, well, worth, the, purch...   
4  [all, of, my, kids, have, cried, non-stop, whe...   

                                          word_count  awesome  great  \
0  {u'and': 1, u'wipes,': 1, u'because': 1, u'ord...        0      0   
1  {u'and': 1, u'love': 1, u'it': 1, u'highly': 1...        0      0   
2  {u'and': 1, u'quilt': 1, u'it': 1, u'comfortab...        0      0   
3  {u'and': 1, u'ingenious': 1, u'love': 1, u'pos...        0      0   
4  {u'and': 1, u'parents!!': 1, u'all': 1, u'help...        0      1   

   fantastic  amazing  love  horrible  bad  terrible  awful  wow  hate  \
0          0        0     0         0    0         0      0    0     0   
1          0        0     1         0    0         0      0    0     0   
2          0        0     0         0    0         0      0    0     0   
3          0        0     1         0    0         0      0    0     0   
4          0        0     0         0    0         0      0    0     0   

   awesome_count  
0              0  
1              0  
2              0  
3              0  
4              0

In [182]:
products.iloc[10761]['awesome']

1

In [183]:
products.iloc[10761]['awesome_count']

1

# Find most used word from selected_words

In [228]:
def selected_word_count():
    count_dict = {}
    for word in selected_words:
        count_dict[word] = products[word].sum()
    return count_dict

In [231]:
print(selected_word_count())

{'fantastic': 908, 'love': 34714, 'bad': 3518, 'awesome': 2017, 'great': 38038, 'terrible': 713, 'amazing': 1311, 'horrible': 697, 'awful': 364, 'hate': 1153, 'wow': 132}


'great': 38038
'wow': 132

# Create a new sentiment analysis model using only selected words

# Add sentiment column

In [271]:
products = products[products['rating'] != 3]

In [274]:
products['sentiment'] = products['rating'] >= 4

## Test train split

In [319]:
from sklearn.cross_validation import train_test_split

In [320]:
train_data, test_data = train_test_split(products, test_size=0.2, random_state = 0)

## Train LogisticRegression model

In [321]:
from sklearn.linear_model import LogisticRegression

In [322]:
selected_words_model = LogisticRegression()

In [323]:
selected_words_model.fit(X=train_data[selected_words], y=train_data['sentiment'])

LogisticRegression(C=1.0, class_weight=None, dual=False, fit_intercept=True,
          intercept_scaling=1, max_iter=100, multi_class='ovr', n_jobs=1,
          penalty='l2', random_state=None, solver='liblinear', tol=0.0001,
          verbose=0, warm_start=False)

## Results

In [324]:
selected_words_model.coef_

array([[ 1.19313429,  1.02912457,  0.78392209,  0.99059312,  1.66812632,
        -2.09732567, -1.038261  , -2.11365292, -1.75002809, -0.10295027,
        -1.53998467]])

In [325]:
selected_words_model.classes_

array([False,  True], dtype=bool)

In [326]:
from itertools import izip as zip
coef_dict = {k: v for k, v in zip(selected_words, selected_words_model.coef_[0])}
print coef_dict

{'fantastic': 0.78392208898133964, 'love': 1.6681263240389248, 'bad': -1.0382610049641134, 'awesome': 1.1931342878557001, 'great': 1.0291245670790194, 'terrible': -2.1136529225594409, 'amazing': 0.9905931183811475, 'horrible': -2.0973256696661782, 'awful': -1.7500280856839332, 'hate': -1.5399846718228309, 'wow': -0.10295027017966574}


# Comparing the accuracy of different sentiment analysis model

In [333]:
selected_words_model.predict(test_data[selected_words])

array([ True,  True,  True, ...,  True,  True,  True], dtype=bool)

In [334]:
test_data_copy = test_data.copy()

In [335]:
pred = pandas.Series(selected_words_model.predict(test_data[selected_words]))

In [336]:
test_data_copy['pred'] = pred.values

In [337]:
test_data_copy.head()

name  \
103733    Luvable Friends Applique Side Closure Bib, Pink   
157975  Infant Optics DXR5 Wall Socket Power Adapter (...   
127693  NUK Active Silicone Spout Learning Cup, Ladybu...   
113718           Luvable Friends 12 Pack Washcloths, Blue   
20416                    Infantino Cart Safari w/Microban   

                                                   review  rating  \
103733  Got this for a friend for her baby shower. Lov...       5   
157975  Ok, so obviously there isn't much to say about...       5   
127693  We have a couple versions of Tommie Tippie and...       5   
113718  Ordered washcloths pictured but others were se...       1   
20416   perfect for what its supposed to be. my 9 mont...       5   

                                             review_split  \
103733  [got, this, for, a, friend, for, her, baby, sh...   
157975  [ok,, so, obviously, there, isn't, much, to, s...   
127693  [we, have, a, couple, versions, of, tommie, ti...   
113718  [ordered, washcloths, pictured, but, others, w...   
20416   [perfect, for, what, its, supposed, to, be., m...   

                                               word_count  awesome  great  \
103733  {u'a': 1, u'loved': 1, u'it!thanks': 1, u'for'...        0      0   
157975  {u'and': 1, u'there': 1, u'wall': 1, u'into': ...        0      0   
127693  {u'and': 1, u'all': 1, u'old': 1, u'nuby,': 1,...        0      0   
113718  {u'them': 1, u'ordered': 1, u'sent.': 1, u'oth...        0      0   
20416   {u'and': 1, u'old': 1, u'extra': 1, u'others':...        0      0   

        fantastic  amazing  love  horrible  bad  terrible  awful  wow  hate  \
103733          0        0     0         0    0         0      0    0     0   
157975          0        0     0         0    0         0      0    0     0   
127693          0        0     0         0    0         0      0    0     0   
113718          0        0     0         0    0         0      0    0     0   
20416           0        0     1         0    0         0      0    0     0   

        awesome_count sentiment  pred  
103733              0      True  True  
157975              0      True  True  
127693              0      True  True  
113718              0     False  True  
20416               0      True  True

In [345]:
num_correct = int(len(test_data_copy[test_data_copy['sentiment'] == test_data_copy['pred']]))

In [346]:
num_correct

28085

In [350]:
total = int(len(test_data_copy))

In [352]:
from __future__ import division
num_correct / total

0.8421036850469251

In [353]:
num_positive = int(len(test_data_copy[test_data_copy['sentiment'] == True]))

In [354]:
num_positive

27989

In [355]:
num_positive / total

0.8392252106383616

# Interpreting the difference in performance between the models

In [369]:
diaper_champ_reviews = products[products['name'] == 'Baby Trend Diaper Champ']

In [370]:
diaper_champ_reviews.describe()

rating  awesome       great   fantastic  amazing        love  \
count  298.000000    298.0  298.000000  298.000000    298.0  298.000000   
mean     4.197987      0.0    0.181208    0.003356      0.0    0.261745   
std      1.332473      0.0    0.385838    0.057928      0.0    0.440324   
min      1.000000      0.0    0.000000    0.000000      0.0    0.000000   
25%      4.000000      0.0    0.000000    0.000000      0.0    0.000000   
50%      5.000000      0.0    0.000000    0.000000      0.0    0.000000   
75%      5.000000      0.0    0.000000    0.000000      0.0    1.000000   
max      5.000000      0.0    1.000000    1.000000      0.0    1.000000   

         horrible         bad    terrible       awful    wow        hate  \
count  298.000000  298.000000  298.000000  298.000000  298.0  298.000000   
mean     0.013423    0.083893    0.003356    0.003356    0.0    0.006711   
std      0.115270    0.277693    0.057928    0.057928    0.0    0.081785   
min      0.000000    0.000000    0.000000    0.000000    0.0    0.000000   
25%      0.000000    0.000000    0.000000    0.000000    0.0    0.000000   
50%      0.000000    0.000000    0.000000    0.000000    0.0    0.000000   
75%      0.000000    0.000000    0.000000    0.000000    0.0    0.000000   
max      1.000000    1.000000    1.000000    1.000000    0.0    1.000000   

       awesome_count  
count          298.0  
mean             0.0  
std              0.0  
min              0.0  
25%              0.0  
50%              0.0  
75%              0.0  
max              0.0

In [371]:
diaper_champ_reviews = diaper_champ_reviews[diaper_champ_reviews['rating'] != 3]

In [372]:
diaper_champ_reviews['sentiment'] = diaper_champ_reviews['rating'] >= 4

In [373]:
diaper_champ_reviews.head()

name  \
312  Baby Trend Diaper Champ   
314  Baby Trend Diaper Champ   
315  Baby Trend Diaper Champ   
316  Baby Trend Diaper Champ   
317  Baby Trend Diaper Champ   

                                                review  rating  \
312  Ok - newsflash.  Diapers are just smelly.  We'...       4   
314  My husband and I selected the Diaper "Champ" m...       1   
315  Excellent diaper disposal unit.  I used it in ...       5   
316  We love our diaper champ.  It is very easy to ...       5   
317  Two girlfriends and two family members put me ...       5   

                                          review_split  \
312  [ok, -, newsflash., diapers, are, just, smelly...   
314  [my, husband, and, i, selected, the, diaper, "...   
315  [excellent, diaper, disposal, unit., i, used, ...   
316  [we, love, our, diaper, champ., it, is, very, ...   
317  [two, girlfriends, and, two, family, members, ...   

                                            word_count  awesome  great  \
312  {u'just': 1, u'less': 1, u'when': 1, u'smell-c...        0      0   
314  {u'things': 1, u'bags,': 1, u'less': 1, u'when...        0      0   
315  {u'control': 1, u'am': 1, u'it': 1, u'nose!!!'...        0      0   
316  {u'and': 1, u'over.': 1, u'all': 1, u'love': 1...        0      0   
317  {u'out,': 1, u'just': 1, u'when': 1, u'both': ...        0      0   

     fantastic  amazing  love  horrible  bad  terrible  awful  wow  hate  \
312          0        0     0         0    0         0      0    0     0   
314          0        0     0         0    0         0      0    0     0   
315          0        0     0         0    0         0      0    0     0   
316          0        0     1         0    0         0      0    0     0   
317          0        0     0         1    0         0      0    0     0   

     awesome_count sentiment  
312              0      True  
314              0     False  
315              0      True  
316              0      True  
317              0      True

In [376]:
pred = pandas.Series(selected_words_model.predict(diaper_champ_reviews[selected_words]))

In [377]:
diaper_champ_reviews['predicted'] = pred.values

In [378]:
diaper_champ_reviews.head()

name  \
312  Baby Trend Diaper Champ   
314  Baby Trend Diaper Champ   
315  Baby Trend Diaper Champ   
316  Baby Trend Diaper Champ   
317  Baby Trend Diaper Champ   

                                                review  rating  \
312  Ok - newsflash.  Diapers are just smelly.  We'...       4   
314  My husband and I selected the Diaper "Champ" m...       1   
315  Excellent diaper disposal unit.  I used it in ...       5   
316  We love our diaper champ.  It is very easy to ...       5   
317  Two girlfriends and two family members put me ...       5   

                                          review_split  \
312  [ok, -, newsflash., diapers, are, just, smelly...   
314  [my, husband, and, i, selected, the, diaper, "...   
315  [excellent, diaper, disposal, unit., i, used, ...   
316  [we, love, our, diaper, champ., it, is, very, ...   
317  [two, girlfriends, and, two, family, members, ...   

                                            word_count  awesome  great  \
312  {u'just': 1, u'less': 1, u'when': 1, u'smell-c...        0      0   
314  {u'things': 1, u'bags,': 1, u'less': 1, u'when...        0      0   
315  {u'control': 1, u'am': 1, u'it': 1, u'nose!!!'...        0      0   
316  {u'and': 1, u'over.': 1, u'all': 1, u'love': 1...        0      0   
317  {u'out,': 1, u'just': 1, u'when': 1, u'both': ...        0      0   

     fantastic  amazing  love  horrible  bad  terrible  awful  wow  hate  \
312          0        0     0         0    0         0      0    0     0   
314          0        0     0         0    0         0      0    0     0   
315          0        0     0         0    0         0      0    0     0   
316          0        0     1         0    0         0      0    0     0   
317          0        0     0         1    0         0      0    0     0   

     awesome_count sentiment predicted  
312              0      True      True  
314              0     False      True  
315              0      True      True  
316              0      True      True  
317              0      True     False

# Use GraphLab

In [379]:
import graphlab

In [ ]:
selected_words_gl = graphlab.logistic_classifier.create(train_data,
                                                    target='sentiment',
                                                    features=[selected_words],
                                                    validation_set=test_data)